### PR TITLE
[ LFX 2026 ] feat(exceptions): evaluate PostureExceptionPolicy.ObjectSelector via labels.Selector

### LINKED ISSUE #2363: [ LFX 2026 ] A `SecurityException` that uses the documented `spec.match.objectSelector` field is rejected wholesale, every posture entry in the CRD is silently dropped
> Part of the SecurityException CRD work tracked in kubescape/kubescape#1982. Current-code gap: a schema-valid match field is unsupported and its presence discards the entire exception.

---

# Description

`objectSelector` is documented as one of the four top-level match fields (alongside `namespaceSelector`, `resources`, `images`) and is present in the CRD schema and design examples:

```yaml
match:
  objectSelector:
    matchLabels:
      app: nginx
```

In the posture getter, encountering `objectSelector` is treated as a hard error:

```go
// core/cautils/getter/crdexceptionsgetter.go:190-201
_, objectSelectorFound, err := unstructured.NestedFieldNoCopy(obj.Object, "spec", "match", "objectSelector")
if err != nil {
	return nil, fmt.Errorf("read objectSelector: %w", err)
}
if objectSelectorFound {
	logger.L().Warning("SecurityException CRD uses unsupported spec.match.objectSelector; skipping", ...)
	return nil, fmt.Errorf("spec.match.objectSelector is not supported")
}
```

That error propagates up to `convertCRDObjectToPosturePolicies` → `GetExceptions`, where the per-CRD failure is swallowed with `continue`:

```go
// core/cautils/getter/crdexceptionsgetter.go:80-86
for i := range seList.Items {
	policies, convErr := convertCRDObjectToPosturePolicies(&seList.Items[i], "SecurityException", g.k8sClient)
	if convErr != nil {
		continue   // <-- the ENTIRE CRD (all posture entries) is dropped
	}
	out = append(out, policies...)
}
```

So a `SecurityException` that combines `objectSelector` with, say, three posture entries does not get "objectSelector ignored, the rest applied" — it gets **nothing** applied, with only a `Warning` line as evidence. The CRD validates at admission (the field is in the schema) but is inert at scan time. That mismatch between "accepted by the API server" and "silently does nothing" is the worst failure mode for a security-suppression resource.

### This is implementable without touching opa-utils

I traced the matching path. `objectSelector.matchLabels` can be supported entirely inside the getter:

- `identifiers.PortalDesignator.DigestAttributesDesignator` routes any **non-reserved** attribute key (anything that isn't `namespace`/`cluster`/`kind`/`name`/`path`/`resourceID`/`k8sResourceHash`) into the `labels` bucket (`armoapi-go/identifiers/designators_methods.go:88-100`).
- `Processor.metadataHasException` then runs `compareLabels` against the workload's labels for those keys (`opa-utils/exceptions/exceptionprocessor.go:266-271`, `comparator.go:60-79`).

So flattening `matchLabels` into each resource designator's `Attributes` map "just works" with the existing comparator — no opa-utils change. Two caveats to handle in the implementation:

1. `compareLabels` matches label values via `regexCompare`, so a label value like `1.25` is treated as a regex. Values must be regex-escaped before being placed in the designator.
2. `matchExpressions` (`In`/`NotIn`/`Exists`/`DoesNotExist`) cannot be represented as key→value attributes and the comparator can't evaluate them. `matchLabels` should be implemented now; `matchExpressions` either resolved at fetch time (list workloads, emit per-resource designators) or documented as best-effort for `v1beta1`.

# Environment

- Repo: `kubescape` at `core/cautils/getter/crdexceptionsgetter.go`
- Affects any `SecurityException`/`ClusterSecurityException` that sets `spec.match.objectSelector`.

# Steps To Reproduce

1. Apply a `SecurityException` with `spec.match.objectSelector.matchLabels` and one or more posture entries.
2. Run an in-cluster scan.
3. Observe none of the posture entries are applied; the only signal is a `Warning: ... unsupported spec.match.objectSelector; skipping`.

# Expected behavior

`objectSelector.matchLabels` constrains the exception to workloads carrying those labels (AND-combined with `resources`/`namespaceSelector` per the design doc's match semantics). The exception's posture entries apply to matching workloads.

# Actual Behavior

The presence of `objectSelector` discards the whole CRD's posture entries; nothing is excepted.

# Suggested fix

In `buildResourceDesignators`, replace the reject-and-error block with conversion:

- Decode `spec.match.objectSelector` into a `metav1.LabelSelector`.
- For `matchLabels`, merge the (regex-escaped) key/value pairs into every resource designator produced for the CRD (and into the namespace-only / fallback designator when `resources` is empty), so the AND semantics hold.
- For `matchExpressions`, resolve at fetch time or document as best-effort for `v1beta1`.

Until `matchExpressions` is fully handled, prefer **partial application** over fail-closed: apply what is supported and emit a single scoped warning, rather than discarding the entire exception.

### Regression test

Converter tests covering: `objectSelector.matchLabels` alone; `objectSelector` + `resources` (AND); regex-escaping of label values; and `matchExpressions` best-effort behavior. Add to `crdexceptionsgetter_test.go`.

# Source

- `core/cautils/getter/crdexceptionsgetter.go:190-201` — `objectSelector` rejected with an error.
- `core/cautils/getter/crdexceptionsgetter.go:80-98` — per-CRD `convErr` swallowed with `continue`, dropping all entries.
- `armoapi-go/identifiers/designators_methods.go:88-100` — `DigestAttributesDesignator` routes non-reserved keys into `labels`.
- `opa-utils/exceptions/exceptionprocessor.go:266-271` and `opa-utils/exceptions/comparator.go:60-79` — `compareLabels` regex-matches designator labels against workload labels.

### PR DESCRIPTION
**Pull request by yugal07**
## Overview

opa-utils is the engine that decides which posture/vulnerability findings a
`PostureExceptionPolicy` suppresses (`exceptions/exceptionprocessor.go`). Its only
workload-matching axis today is `Resources []identifiers.PortalDesignator`, where each
designator encodes match values as **per-key anchored regexes** (`compareLabels` →
`regexCompare`). That path can express *"label key exists and its value matches this
regex"* — a positive, per-key AND — but it **cannot** express the set-based label
operators `NotIn` / `DoesNotExist` (RE2 has no negative lookahead and the comparator
requires the key to be present).

The SecurityException CRD exposes a native Kubernetes `spec.match.objectSelector`
(`matchLabels` **and** `matchExpressions`). [kubescape/kubescape#2366](https://github.com/kubescape/kubescape/pull/2366) honors
`matchLabels` today by flattening it into regex designators and ships `matchExpressions`
as a **logged, best-effort no-op**. [armosec/armoapi-go#665](https://github.com/armosec/armoapi-go/pull/665) added the typed
`PostureExceptionPolicy.ObjectSelector` field so a full selector can travel end-to-end.

This PR makes opa-utils **consume** that field: when an exception carries an
`ObjectSelector`, it is evaluated with `labels.Selector.Matches(...)` semantics against
the workload's labels, and ANDed with the designator match. This is step 2 of the
three-PR chain and unblocks retiring the `matchExpressions` best-effort skip.

## What changed

- **`parseObjectSelector`** converts `exception.ObjectSelector` into a `labels.Selector`
  once, with an explicit nil/empty/malformed contract (below), and returns `(nil, true)`
  when there is no label constraint so callers skip the label axis.
- The parsed selector is **threaded down** `hasException → iterateRegoResponseVector →
  metadataHasException` and evaluated inside `metadataHasException` against **the exact
  object the designator is checked against** — for a `RegoResponseVector` that is each
  *related* object, not the label-less envelope.
- The label match is **labels only** (not annotations), per the SecurityException spec
  wording ("selects workloads by their labels") and deliberately unlike the OR
  labels/annotations behavior of the regex-designator path.

### Why the selector is evaluated per related object (the correctness point)

An earlier cut evaluated the selector once against the top-level workload object. For a
`RegoResponseVector` the envelope carries **no labels** (they live in the related
objects), so any non-empty `objectSelector` would never match — findings on
vector-wrapped workloads could never be excepted — and the selector and the designator
could be satisfied by *different* related objects (a cross-object false match). Threading
the selector into `metadataHasException` fixes both: a single object must satisfy **both**
the designator and the selector for the exception to apply. This is pinned by
`TestGetResourceExceptions_ObjectSelector_RegoResponseVector`, including the
cross-object case.

## Semantics / contract

- **Nil-or-empty selector → no label constraint** (the label axis is skipped). This is a
  deliberate inverse of `metav1.LabelSelectorAsSelector`, whose conversion disagrees with
  the intent: `nil` yields `labels.Nothing()` (would silently disable every
  selector-less exception — the common case) and an empty selector yields
  `labels.Everything()`. Both are guarded before converting.
- **Malformed selector → matches nothing**, never degrades into match-all. Unreachable
  from a validated CRD and unset on cloud exceptions, so it is treated as a defensive
  match-nothing rather than logged here.
- **ANDed with the designator, not standalone.** `ObjectSelector` is evaluated *inside*
  the per-designator `hasException`, which early-returns when a designator's attributes
  are all empty. So the exception still needs at least one non-empty designator for the
  selector to be reached — which the kubescape adapter guarantees: it always emits a
  fallback designator (`{namespace}` for namespaced, `{kind:"*"}` for cluster-scoped)
  and ANDs `objectSelector` into it ([kubescape/kubescape#2366](https://github.com/kubescape/kubescape/pull/2366)). Consumers pairing this
  release with the adapter should populate `ObjectSelector` and may drop the now-redundant
  regex-escaped `matchLabels`-in-designators.

## Dependencies

The `feat` commit bumps `armosec/armoapi-go` `v0.0.330 → v0.0.730` to pick up the
`ObjectSelector` field ([armosec/armoapi-go#665](https://github.com/armosec/armoapi-go/pull/665)), which also moves
`k8s.io/{api,apimachinery,client-go}` `v0.29.2 → v0.32.0`. `k8s.io/apimachinery`
(`labels`, `metav1`) is already a direct dependency of the module, so no genuinely new
dependency is introduced — but the go.mod/go.sum churn is intentional and required for
the typed field.

## How to Test

```bash
go test ./exceptions/...   # ok
go vet ./exceptions/...     # clean
```

New/extended unit tests in `exceptions/exceptionprocessor_test.go`:

| Test | What it pins down |
| --- | --- |
| `TestGetResourceExceptions_ObjectSelector` | matchLabels; `In`/`NotIn`/`Exists`/`DoesNotExist`; nil/empty impose no constraint; non-empty vs a label-less workload; malformed matches nothing |
| `TestGetResourceExceptions_ObjectSelector_RegoResponseVector` | selector is evaluated against the *related* workload of a vector; selector + designator must be satisfied by the *same* related object (no cross-object match) |
| `TestGetResourceExceptions_ObjectSelectorAndContainerName` | selector AND `containerName` designator are both required; a selector miss short-circuits before the container check |
| `TestHasException_RegoResponseVector_BaseFallthroughWithSelector` | the base-object fall-through is fail-closed under a non-empty selector: a related-object label match does not cross-combine with a base-envelope designator match |
| `TestHasException`, `TestProcessor_iterateRegoResponseVector`, `TestMetadataHasException_ContainerName`, container-precision tests | updated for the new `selector` parameter (passed `nil`) — existing behavior unchanged when no selector is present |

## Related issues/PRs

- Part of [kubescape/kubescape#1982](https://github.com/kubescape/kubescape/issues/1982) — SecurityException CRD implementation (epic).
- Consumes [armosec/armoapi-go#665](https://github.com/armosec/armoapi-go/pull/665) — the typed `PostureExceptionPolicy.ObjectSelector` field.
- Builds on [kubescape/kubescape#2366](https://github.com/kubescape/kubescape/pull/2366) — honors `objectSelector.matchLabels` and ships
  `matchExpressions` as a logged best-effort no-op; this PR is the evaluation half that
  lets the follow-up kubescape PR retire that skip and close [kubescape/kubescape#2363](https://github.com/kubescape/kubescape/issues/2363).
- Companion CRD shipping: [kubescape/helm-charts#859](https://github.com/kubescape/helm-charts/pull/859) —
  adds CEL validation to the `SecurityException` / `ClusterSecurityException` CRDs (`v1beta1`) whose
  `spec.match.objectSelector` this PR evaluates.
- Design reference: [`docs/security-exception-design.md`](https://github.com/kubescape/kubevuln/blob/main/docs/security-exception-design.md) (`kubescape/kubevuln`, `v1beta1`).

## Checklist before requesting a review

- [x] My code follows the style guidelines of this project
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have performed a self-review of my code
- [x] I have added tests covering the new axis (matchLabels, matchExpressions, nil/empty/malformed, and the RegoResponseVector per-object + cross-object cases)
- [x] New and existing unit tests pass locally with my changes

## Summary by CodeRabbit

* **New Features**
  * Exception policies now support label-based `ObjectSelector`, applying exceptions only to matching workloads.
  * Exception matching now combines designator constraints with object-label selector checks for tighter control.
* **Bug Fixes**
  * Malformed label selectors are ignored to prevent unintended behavior.
  * Selector evaluation is applied per related object and can fail closed to avoid cross-object matches.
* **Chores**
  * Updated Go toolchain and refreshed project dependencies.
  * Adjusted CI Go setup to derive the Go version from `go.mod`.
  * Updated test error reporting for clearer failures.

### CODE REVIEW

**Comment by yugal07** [comment]
@matthyx The CI seems to be failing on this PR

  What's failing: the Test units step errors with go: no such tool "covdata". It's not a code issue — all tests pass. It's a Go
  toolchain-resolution problem.

  Root cause: the CodeRabbit review asked to bump go.opentelemetry.io/otel/sdk to >= v1.43.0 (fixes CVE-2026-24051 /
  GHSA-9h8m-3fm2-qjrq, PATH hijacking). The otel advisory has no patched release below v1.43.0, and every otel v1.43.0 module
  declares go 1.25.0. Since Go 1.21, go get/go mod tidy raises the main module's go directive to satisfy dependencies, so
  go.mod moved go 1.24.1 → go 1.25.0 (and dropped the toolchain go1.24.2 line).

  CI pins go-version: '1.21' and runs go test -covermode=count -coverprofile. With go 1.25.0 and no toolchain line, the
  auto-switch to 1.25 breaks the coverage step (covdata). Previous green runs had go 1.24.1 / toolchain go1.24.2, which
  switched cleanly.

  So the otel fix and the Go-1.21 CI are in direct tension — taking the otel patch necessarily requires Go 1.25.

  What should i do? Some solutions could be 
  1. Bump CI to the module's Go version (recommended): switch pr.yaml + release.yaml to go-version-file: go.mod (or '1.25') so
  the runner installs 1.25 directly. Honest fix — the module genuinely needs 1.25 now. Touches workflow files.
  2. toolchain line only: add toolchain go1.25.0 back to go.mod, no CI change. Minimal, mirrors the previous explicit-toolchain
  pattern, but I can't fully validate the covdata re-exec without Go 1.21 locally.
  3. Drop the otel bump: keep the grpc (v1.79.3) and golang-jwt/jwt/v4 (v4.5.1) fixes, revert otel/sdk to v1.37.0. Keeps go 
  1.24.1 and CI unchanged, but forgoes the otel fix (a deep transitive dep; the advisory is a macOS/BSD PATH-hijack, low
  relevance to this library).

  Happy to push whichever you'd like. If you're fine standardizing CI on go-version-file: go.mod, I'll go with option 1.

**Comment by matthyx** [comment]
@yugal07 yeah bump to 1.25 (solution 1)

File: exceptions/exceptionprocessor.go [hunk 1]
```diff
@@ -223,13 +267,16 @@ func (p *Processor) hasException(clusterName string, designator *identifiers.Por
 		if _, ok := attributes.GetLabels()[identifiers.AttributeContainerName]; ok {
 			return false
 		}
-		// otherwise, continue to check the base object
+		// otherwise, continue to check the base object. A non-empty objectSelector
+		// will not match the label-less base envelope, so an exception whose selector
+		// matched only a related object (never the base) is correctly not applied here
+		// — consistent with the same-object AND enforced in metadataHasException.
```
Comment (matthyx) [inline exceptions/exceptionprocessor.go]: **Blocker — negative `objectSelector` operators (`DoesNotExist` / `NotIn`) wrongly match the label-less RegoResponseVector base envelope on fall-through, producing a false exception (a security finding is suppressed when it should not be).**

The comment here asserts the fall-through is fail-closed: *"A non-empty objectSelector will not match the label-less base envelope."* That holds only for **positive** selectors. A negative requirement (`env DoesNotExist`, `env NotIn [prod]`) matches an **empty** label set, so when related-object iteration fails and control falls through to `metadataHasException(<base envelope>, …, selector)`, the selector check at line ~306 passes against the envelope's empty labels. If the designator also matches the envelope (e.g. a name-based designator, or a `{namespace}` designator when the vector carries a namespace), the exception is applied even though **no real workload satisfied the selector**.

Reproduced against this branch (all tests otherwise green):

```go
// vector "web"; related Deployment HAS env=prod  ->  must NOT be excepted by "env DoesNotExist"
designator := {name: "web"}
objectSelector := {matchExpressions: [{key: env, operator: DoesNotExist}]}
GetResourceExceptions(...) => 1   // WANT 0  — finding on a workload with env=prod is wrongly suppressed
```

`TestHasException_RegoResponseVector_BaseFallthroughWithSelector` only covers a *positive* `matchLabels` selector, so this path is untested.

**Suggested fix:** the base envelope is not a real labeled workload, so the label axis should never be evaluated against it. When a non-empty selector is present, skip the base fall-through entirely for a RegoResponseVector:

```go
if _, ok := attributes.GetLabels()[identifiers.AttributeContainerName]; ok {
    return false
}
if selector != nil {
    return false // selector is a workload-label axis; the label-less base envelope is not a workload
}
// otherwise, continue to check the base object
```

Please also add a regression test with a negative operator (`DoesNotExist` / `NotIn`) over a vector to pin the fail-closed behavior.

**Review by matthyx — APPROVED** [review]
